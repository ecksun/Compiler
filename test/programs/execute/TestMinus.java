class TestMinus {
    public static void main(String[] args) {
        if (new TestMinusHelper().testMinus()) {} else {}
    }
}

class TestMinusHelper {
    public boolean testMinus() {
        int x;

        // NOTE: All integers operations are well-defined mod 2^32.
        x = this.print(2-1); // subtract the constant 1
        x = this.print(2+-1); // add the contant -1
        x = this.print(2--1); // subtract the constant -1
        x = this.print(2-2147483647); // subtract largest positive int
        x = this.print(2+-2147483648); // add largest negative int
        x = this.print(2--2147483648); // subtract largest negative int

        // Correct Java for reference output.
        //x = print(2-1); // subtract the constant 1
        //x = print(2+(-1)); // add the contant -1
        //x = print(2-(-1)); // subtract the constant -1
        //x = print(2-2147483647); // subtract largest positive int
        //x = print(2+(-2147483648)); // add largest negative int
        //x = print(2-(-2147483648)); // subtract largest negative int

        return true;
    }

    public int print(int x) {
        System.out.println(x);
        return x;
    }
}
