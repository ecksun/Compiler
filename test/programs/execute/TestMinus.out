1
1
3
-2147483645
-2147483646
-2147483646
